src/v/cloud_storage/materialized_resources.cc

```diff
@@ -291,11 +367,11 @@ void materialized_resources::trim_segments(std::optional<size_t> target_free) {
         auto deadline = st.atime + stm_max_idle_time;
 
         // We freed enough, drop out of iterating over segments
-        if (target_free && _segment_units.current() >= target_free) {
+        if (target_free && _mem_units.current() >= target_free) {
             break;
         }
 
-        if (target_free || (now >= deadline)) {
+        if (target_free || st.overcommit || (now >= deadline)) {
```

Comment:
I'm a bit concerned about the fairness of this. Feels like we should trim the oldest segments if we overcommit and not the one we've just added.